Book/php7/extensions_design/hooks.rst

```diff
@@ -0,0 +1,172 @@
+        zend_execute_internal = my_execute_internal;
+
+        original_zend_execute_ex = zend_execute_ex;
+        zend_execute_ex = my_execute_ex;
+
+        return SUCCESS;
+    }
+
+    PHP_MSHUTDOWN_FUNCTION(my_extension)
+    {
+        zend_execute_internal = original_zend_execute_internal;
+        zend_execute_ex = original_zend_execute_ex;
+
+        return SUCCESS;
+    }
+
+One downside of overwriting ``zend_execute_ex`` is that it changes the Zend
+Virtual Machine runtime behavior to use recursion instead of a hybrid mode that
+uses goto/switch. In addition a PHP engine without overwritten
+``zend_execute_ex`` can also generate more optimized function call opcodes.
+
+These hooks are very performance sensitive depending on the complexity of code
+that wraps the original functions.
+
+Overwriting an Internal Function
+********************************
+
+While overwriting the execute hooks an extension can record **every** function
+call, you can also overwrite individual function pointers of userland core and
+extension functions (and methods). This has much better performance
+characteristics if an extension only needs access to specific internal function
+calls.::
+
+    #if PHP_VERSION_ID >= 70300
+    #define EXT_FASTCALL ZEND_FASTCALL
+    #else
+    #define EXT_FASTCALL
+    #endif
+
+    #if PHP_VERSION_ID < 70200
+    typedef void (*zif_handler)(INTERNAL_FUNCTION_PARAMETERS);
+    #endif
+    zif_handler original_handler_var_dump;
+
+    void EXT_FASTCALL my_overwrite_var_dump(INTERNAL_FUNCTION_PARAMETERS)
+    {
+        // if we want to call the original function
+        original_handler_var_dump(INTERNAL_FUNCTION_PARAM_PASSTHRU);
+    }
+
+    PHP_MINIT_FUNCTION(my_extension)
+    {
+        zend_function *original;
+
+        original = zend_hash_str_find_ptr(EG(function_table), "var_dump", sizeof("var_dump")-1);
+
+        if (original != NULL) {
+            original_handler_var_dump = original->internal_function.handler;
+            original->internal_function.handler = my_overwrite_var_dump;
+        }
+    }
+
+When overwriting a class method, then the function table can be found on the
+``zend_class_entry``.::
+
+    zend_class_entry *ce = zend_hash_str_find_ptr(CG(class_table), "PDO", sizeof("PDO")-1);
+    if (ce != NULL) {
+        original = zend_hash_str_find_ptr(&ce->function_table, "exec", sizeof("exec")-1);
+
+        if (original != NULL) {
+            original_handler_pdo_exec = original->internal_function.handler;
+            original->internal_function.handler = my_overwrite_pdo_exec;
+        }
+    }
+
+Modifying the Abstract Syntax Tree (AST)
+****************************************
+
+When PHP 7 compiles PHP code it converts it into an abstract syntax tree (AST)
+before finally generating Opcodes that are persisted in Opcache. One lesser
+known hook is called ``zend_ast_process`` for every compiled script and allows
+you to modify the AST after it is parsed and created.
```

Comment:
Something is missing between "known hook is called zend_ast_process" and "for every compiled script".

Maybe just "The zend_ast_process hook is called for every compiled script".